test/06-device_log.ts

```diff
@@ -0,0 +1,191 @@
+		const fx = await fixtures.load('create-device-log');
+		ctx.loadedFixtures = fx;
+		ctx.user = fx.users.admin;
+		ctx.device = fx.devices.device1;
+		ctx.device2 = fx.devices.device2;
+
+		const res = await supertest(app, ctx.user)
+			.post(`/api-key/device/${ctx.device.id}/device-key`)
+			.expect(200);
+
+		expect(res.body).to.be.a('string');
+		ctx.device.apiKey = res.body;
+
+		const res2 = await supertest(app, ctx.user)
+			.post('/api-key/user/full')
+			.send({ name: 'name' })
+			.expect(200);
+
+		expect(res2.body).to.be.a('string');
+		ctx.user.apiKey = res2.body;
+	});
+	after(() => fixtures.clean(ctx.loadedFixtures));
+
+	const createLog = (extra = {}) => {
+		return {
+			isStdErr: true,
+			isSystem: true,
+			message: 'a log line',
+			timestamp: Date.now(),
+			...extra,
+		};
+	};
+
+	// Creating
+
+	it('should allow devices to write their own logs', () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog()])
+			.expect(201));
+
+	it('should allow devices to write many logs in a batch', () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog(), createLog()])
+			.expect(201));
+
+	it("should not allow devices to write other devices' logs", () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device2.uuid}/logs`)
+			.send([createLog()])
+			// Yields 404 because fetching device2 by uuid yields nothing
+			.expect(404));
+
+	it('should accept and store batches where some logs have serviceId', () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog({ serviceId: 10 }), createLog()])
+			.expect(201));
+
+	it('should return 201 even when the list of logs is empty', () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([])
+			.expect(201));
+
+	it('should ignore logs with uuid (from dependent devices)', () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog({ uuid: 'abcdef' })])
+			.expect(201));
+
+	it('should ignore unknown properties on logs and store them', () =>
+		supertest(app, ctx.device.apiKey)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog({ hello: 'hi there' })])
+			.expect(201));
+
+	it('should reject batches with broken logs', async () => {
+		for (const extra of [
+			{ message: 123 },
+			{ message: null },
+			{ timestamp: new Date().toISOString() },
+		]) {
+			await supertest(app, ctx.device.apiKey)
+				.post(`/device/v2/${ctx.device.uuid}/logs`)
+				.send([createLog(extra)])
+				.expect(400);
+		}
+	});
+
+	// TODO: Move to balena-api test with restricted user, admin user of open-balena-api CAN write log
+	it.skip('should not allow users to write device logs', () =>
+		supertest(app, ctx.user)
+			.post(`/device/v2/${ctx.device.uuid}/logs`)
+			.send([createLog()])
+			.expect(401));
```

Comment:
Should this be removed altogether from here as it can never be tested here?

Reply:
removed